### ISSUE TITLE
Rest Client Metrics

### ISSUE THREAD

**Issue by napstr** [Discord4J/Discord4J#866]
**Feature Description:** 

I'd like to be able to collect metrics about the sent discord rest requests.

Netty's http client has metrics support, but it only offers the full URI, which would lead to a high cardinality, requiring manually making templates out of those again, I'd like to avoid that.
D4J already has an ideal data structure holding the relevant information **http method + uri template**, the `Route` class.
Also the `RequestStream` already collects information about the **response times** of the API.

Additionally, the response codes, both **http status codes** as well as [Discord's **JSON error codes**](https://discord.com/developers/docs/topics/opcodes-and-status-codes#json) where available, would be helpful.

To sum it up:
I'd like to be able to subscribe to a stream of objects that provide the information mentioned above:
- http method
- uri template
- response time
- http status code
- if available: discord JSON error code

What would be a good place to add this functionality? I'm open to providing a pull request adding this functionality. I've spent some time looking through the D4J rest.request package, however I didn't spot a good place. Being new to reactive I feel like I might be missing the obvious solution to this. Help appreciated!

**Justification:** 

Bots need to be able to instrument their behaviour against the Discord API for troubleshooting.

**Comment by danthonywalker** [comment]
Shouldn't logs from io.netty be sufficient for this? Aggregate the logs to something like Elasticsearch and you can query whatever data you need.

**Comment by napstr** [comment]
That sound like a solution, but a rather specific one.
In my context, I collect the metrics in Prometheus, not Elasticsearch, and would prefer not to have to add an additional service just for one specific case.

**Comment by napstr** [comment]
Actually, I think I might have already most if the data availably be providing a custom router. What I am missing is somehow grabbing the response time:


```java
/**
 * Collect metrics about the executed web requests against the Discord API
 */
public class InstrumentedRouter implements Router {

	private final Router delegate;
	private final ReactorResources reactorResources;

	public InstrumentedRouter(Router delegate, RouterOptions routerOptions) {
		this.delegate = delegate;
		this.reactorResources = routerOptions.getReactorResources();
	}

	@Override
	public DiscordWebResponse exchange(DiscordWebRequest request) {
		DiscordWebResponse exchange = delegate.exchange(request);

		Mono<ClientResponse> instrumentedResponse = exchange.mono()
				.doOnNext(response -> instrumentSuccess(response, request))
				.doOnError(throwable -> instrumentError(throwable, request));

		return new DiscordWebResponse(instrumentedResponse, reactorResources);
	}

	private void instrumentSuccess(ClientResponse response, DiscordWebRequest request) {
		HttpResponseStatus status = response.getHttpResponse().status();
		HttpMethod method = request.getRoute().getMethod();
		String uriTemplate = request.getRoute().getUriTemplate();

		// Instrument
	}

	private void instrumentError(Throwable error, DiscordWebRequest request) {
		if (error instanceof ClientException) {
			ClientException clientException = (ClientException) error;

			HttpResponseStatus status = clientException.getStatus();
			HttpMethod method = request.getRoute().getMethod();
			String uriTemplate = request.getRoute().getUriTemplate();

			Optional<ErrorResponse> errorResponseOpt = clientException.getErrorResponse();
			Optional<Map<String, Object>> errorReponseContent = errorResponseOpt.map(ErrorResponse::getFields);

			var message = errorReponseContent.map(fields -> fields.get("message")).orElse(null);
			var code = errorReponseContent.map(fields -> fields.get("code")).orElse(-1);

			// Instrument
		}
	}
}
```

**Comment by darichey** [comment]
Could you get it the same [way we do in `RequestStream`](https://github.com/Discord4J/Discord4J/blob/master/rest/src/main/java/discord4j/rest/request/RequestStream.java#L121-L123)? You can get the `HttpClientResponse` from `ClientResponse`.

**Comment by napstr** [comment]
That's a decent idea, I'm giving that a try right now, accessing the `requestTimestamp` from the reactive context via the signal received on `.doOnEach`.

I think that'll cover 80%. If I understand the structure of D4J correctly, instrumenting the `requestTimestamp` in the Router will not instrument the Discord API response time, but also include the request potentially being stuck in the ratelimiter.

While that is an interesting metric as well, the raw request time is what I'm going for. It should be trivial however to add another timestamp to the context when the response completes, so if I can verify that can use the reactive context, I'll probably provide a PR to add another timestamp.

**Comment by napstr** [comment]
Ah nvm now that I think about it, the `requestTimestamp` is created after the request is past the ratelimiter, right? That would cover my usecase completely indeed.

**Comment by napstr** [comment]
With #867 merged in, I have all the necessary data available now. For anyone else stumbling upon this in the future, this is my implementation: https://github.com/napstr/prometheus_extensions/blob/master/src/main/java/space/npstr/prometheus_extensions/d4j/InstrumentedRouter.java#L52
